Windows arm64 port - Hello world works!

ragavsachdeva opened the pull request:
**Description:**
- First milestone towards porting Go to win/arm64.
- Added the required configs and files to support building Go executables for the target os/arch pair. 
- Implemented a few assembly functions that interface with Go runtime and the operating system.

**Result:**
- Able to build and run helloworld.go for windows/arm64 (and other simple programs).

**Known shortcomings:**
- Still need to provide exception handling support.
- Relocation support is unimplemented (and we have to trick the system into allowing absolute addresses).

**Notes:**
- Most of this work is adapted from win/arm32 port.

Commit:
Added required configs and files to support the win/arm64 build.

Go build toolchain had no support for win/arm64 and threw unsupported
GOOS/GOARCH pair errors every time you'd try to build a program for
this target platform. This commit fills in a lot of those missing
configs and file dependencies in order to successfully allow win/arm64
executables to be built for Go programs. Stubs added are heavily
influenced from win/arm32 files (plus other architectures) but
of course are adapted to arm64 platform architecture (while adhering to
Windows arm64 ABI and Windows PE format documentation).

Commit:
Implemented asmstdcall function.

Given a function pointer and some arguments, asmstdcall function sets up
the stack in accordance with arm64 function calling conventions and
calls the given function.

Commit:
Implemented tstart_stdcall, init_thread_tls and save_g.

Added functionality for windows to start new os thread. The
implementation of functions is directly adapted from arm32 equivalent
with appropriate changes to account architectural and instructional
changes.

Commit:
Added thread local storage support for win/arm64.

During runtime, thread local storage needs to be allocated before it can
be initialised by the os func call. This required changing Go arm64
assembly runtime pipeline a bit to allow for TLS slot allocation.

Commit:
Implemented onosstack and switchtothread.

No clear documentation on how these functions fit into the pipeline.
Adapted directly from arm32 with appropriate instructional changes
to allow for compatibility with arm64.

Commit:
Implemented nanotime and usleep2.

Commit:
Helloworld works!

Added unimplemented placeholder functions that were required for arm32
and will likely be needed for arm64. These functions currently have
breakpoints for debugging purposes.

hbeberman commented on src/cmd/api/goapi.go:
```diff
@@ -80,6 +80,7 @@ var contexts = []*build.Context{
 	{GOOS: "openbsd", GOARCH: "386"},
 	{GOOS: "openbsd", GOARCH: "amd64", CgoEnabled: true},
 	{GOOS: "openbsd", GOARCH: "amd64"},
+	{GOOS: "windows", GOARCH: "arm64"},
```
This list is sorted by OS, move this line next to the other entries.

ragavsachdeva replied:
Actually, I am not convinced this is even needed. Removing it from the pr.

dmcilvaney commented on src/cmd/api/goapi.go:
```diff
@@ -80,6 +80,7 @@ var contexts = []*build.Context{
 	{GOOS: "openbsd", GOARCH: "386"},
 	{GOOS: "openbsd", GOARCH: "amd64", CgoEnabled: true},
 	{GOOS: "openbsd", GOARCH: "amd64"},
+	{GOOS: "windows", GOARCH: "arm64"},
```
Looks like the tuples tend to be grouped by OS, we should follow that convention. Also, is there a reason Windows + arm isn't here?

ragavsachdeva replied:
Actually, I am not convinced this is even needed. Removing it from the pr.

dmcilvaney commented on src/cmd/link/internal/arm64/obj.go:
```diff
@@ -125,6 +126,10 @@ func archinit(ctxt *ld.Link) {
 			*ld.FlagRound = 0x10000
 		}
+
+	case objabi.Hwindows: /* PE executable */
```
Should this have a TODO: or something so we can find it later, or is this simply un-needed for PECOFF?

ragavsachdeva replied:
I don't think anything more is needed here. This case exists just to prevent the function from exiting due to "unknown" ctxt.HeadType.

hbeberman commented on src/cmd/link/internal/arm64/obj.go:
```diff
@@ -125,6 +126,10 @@ func archinit(ctxt *ld.Link) {
 		}
+
+	case objabi.Hwindows: /* PE executable */
+		// ld.HEADR, ld.FlagTextAddr, ld.FlagDataAddr and ld.FlagRound are set in ld.Peinit
```
Just curious, have these already been set by now?
Why is this the only case with a return statement, do we need to skip the if statement at the end of this function?

ragavsachdeva replied:
We need to skip the default case (that throws an error due to "unknown" ctxt.HeadType) at the top of the switch statement.

hbeberman replied:
Understood but do we need to be bypassing the printf warning below by returning in this case? No other case returns.

hbeberman commented on src/cmd/link/internal/ld/pe.go:
```diff
@@ -1372,6 +1387,9 @@ func (rt *peBaseRelocTable) addentry(ctxt *Link, s *sym.Symbol, r *sym.Reloc) {
 	case 4:
 		e.typeOff |= uint16(IMAGE_REL_BASED_HIGHLOW << 12)
+	case 8:
+		// Todo(ragav): This case is a placeholder for now until we add relocation support.
```
Generally don't like seeing Todo in code, is there a corresponding task tracking this?

Google [style guidelines](https://google.github.io/styleguide/cppguide.html#TODO_Comments) say it should be TODO in all caps.

ragavsachdeva replied:
Yes, getting RELOC to work is on the list.

hbeberman commented on src/cmd/link/internal/ld/pe.go:
```diff
@@ -1426,11 +1444,12 @@ func addPEBaseRelocSym(ctxt *Link, s *sym.Symbol, rt *peBaseRelocTable) {
 	switch ctxt.Arch.Family {
 	default:
 		return
+	// Todo(ragav): Need to add Arm64 case here when we provide relocation support.
```
```suggestion
	// TODO(ragav): Need to add ARM64 case here when we provide relocation support.
```

hbeberman commented on src/cmd/link/internal/loadpe/ldpe.go:
```diff
@@ -304,6 +322,12 @@ func Load(arch *sys.Arch, syms *sym.Symbols, input *bio.Reader, pkg string, leng
 					rp.Add = int64(int32(binary.LittleEndian.Uint32(sectdata[rsect][rp.Off:])))
 				}
+
+			// Todo(ragav): add a case for ARM64
```
```suggestion
			// TODO(ragav): add a case for ARM64
```

hbeberman commented on src/cmd/vendor/golang.org/x/sys/windows/asm_windows_arm64.s:
```diff
@@ -0,0 +1,11 @@
+// Copyright 2018 The Go Authors. All rights reserved.
```
```suggestion
// Copyright 2019 The Go Authors. All rights reserved.
```

hbeberman replied:
Fix copyrights in each new file to 2019.

hbeberman commented on src/runtime/asm_arm64.s:
```diff
@@ -15,17 +15,28 @@ TEXT runtime·rt0_go(SB),NOSPLIT,$0
 
-	// create istack out of the given (operating system) stack.
-	// _cgo_init may update stackguard.
+	
```
nit: Extra whitespace.

dmcilvaney commented on src/runtime/asm_arm64.s:
```diff
@@ -15,17 +15,28 @@ TEXT runtime·rt0_go(SB),NOSPLIT,$0
 
-	// create istack out of the given (operating system) stack.
-	// _cgo_init may update stackguard.
+	
```
Is there an easy way to regression test this? i.e. make sure this doesn't break ARM64 Linux.

ragavsachdeva replied:
I don't believe it should be a problem but it's on my list. I will look into it at some point.

hbeberman commented on src/runtime/asm_arm64.s:
```diff
@@ -1100,10 +1104,14 @@ TEXT _cgo_topofstack(SB),NOSPLIT,$24
+		// This only happens if iscgo, so jump straight to save_g
+		BL	runtime·save_g(SB)
+		RET
+//	#endif
```
Don't leave commented out code.

ragavsachdeva replied:
Added a todo.

hbeberman commented on src/runtime/defs_windows_arm64.go:
```diff
@@ -0,0 +1,151 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Todo(ragav): adapt to arm64
```
```suggestion
// TODO(ragav): adapt to arm64
```

dmcilvaney commented on src/runtime/asm_arm64.s:
```diff
@@ -1100,10 +1104,14 @@ TEXT _cgo_topofstack(SB),NOSPLIT,$24
-	// This only happens if iscgo, so jump straight to save_g
-	BL	runtime·save_g(SB)
-	RET
+//	#ifdef GOOS_windows
```
Either clean up or add TODO?

ragavsachdeva replied:
Added a TODO to clean up.

hbeberman commented on src/runtime/defs_windows_arm64.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+func stackcheck() {
+	// TODO: not implemented on ARM
```
```suggestion
	// TODO(ragav): not implemented on ARM64
```

hbeberman commented on src/runtime/sys_windows_arm64.s:
```diff
@@ -0,0 +1,501 @@
+	MOVD	LR, R21			// R21 = link register
+	
+	// SetLastError(0)
+	WORD	$0xaa1203e1		// MOVD	R18, R1
```
Why can't this be expressed in assembly?

ragavsachdeva replied:
R18 is a reserved register in arm64 and Go compiler doesn't let you use it. But we need it as the base address of TEB structure is stored here (as per [windows arm64 ABI](https://docs.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions?view=vs-2019)).

dmcilvaney replied:
Regardless of correctness, it looks very out of place. A comment explaining why this was needed would help future devs.

hbeberman commented on src/runtime/sys_windows_arm64.s:
```diff
@@ -0,0 +1,501 @@
+
+	// Save return values
+	MOVD	R0, 24(R19)
+	MOVD	R1, 32(R19)		// Note(ragav): R1 is not a result register in ARM64. Double check this line.
```
TODO?

ragavsachdeva replied:
Sure.

hbeberman commented on src/runtime/sys_windows_arm64.s:
```diff
@@ -0,0 +1,501 @@
+//   +----------------+
+// 24| argument (r0)  |
+//---+----------------+
+// 16 | param1         |
```
```suggestion
// 16| param1         |
```
Fix diagram alignment

hbeberman commented on src/runtime/sys_windows_arm64.s:
```diff
@@ -0,0 +1,501 @@
+//   +----------------+
+//
+TEXT runtime·externalthreadhandler(SB),NOFRAME,$0
+	// Note(ragav): Check if the function needs to be nosplit.
```
```suggestion
	// TODO(ragav): Check if the function needs to be nosplit.
```

hbeberman commented on src/runtime/sys_windows_arm64.s:
```diff
@@ -0,0 +1,501 @@
+
+// uint32 tstart_stdcall(M *newm);
+TEXT runtime·tstart_stdcall(SB),NOSPLIT|NOFRAME,$0
+ 	// Todo(ragav): save non-volatile registers, if needed
```
```suggestion
 	// TODO(ragav): save non-volatile registers, if needed
```

hbeberman commented on src/runtime/syscall_windows.go:
```diff
@@ -49,6 +52,9 @@ func callbackasmAddr(i int) uintptr {
 		// On ARM, each entry is a MOV instruction
 		// followed by a branch instruction
 		entrySize = 8
+	//todo(ragav): check for the correctness of arm64 case
```
```suggestion
	//TODO(ragav): check for the correctness of arm64 case
```

hbeberman commented on src/runtime/tls_arm64.s:
```diff
@@ -8,12 +8,22 @@
 #include "textflag.h"
 #include "tls_arm64.h"
 
+//todo(ragav): add support for windows
```
```suggestion
//TODO(ragav): add support for windows
```

hbeberman commented on src/runtime/tls_arm64.s:
```diff
@@ -8,12 +8,22 @@
 
+//todo(ragav): add support for windows
 TEXT runtime·load_g(SB),NOSPLIT,$0
+	// Todo(ragav): Following breakpoint is for debugging purposes only.
```
```suggestion
	// TODO(ragav): Following breakpoint is for debugging purposes only.
```

hbeberman commented on src/runtime/tls_arm64.s:
```diff
@@ -8,12 +8,22 @@
 	BEQ	nocgo
 
+#ifdef GOOS_windows
+	WORD	$0xaa1203e0		// MOVD	R18, R0
```
Does "MOVD	R18, R0" not assemble?

ragavsachdeva replied:
More like it doesn't compile.

hbeberman commented on src/runtime/wincallback.go:
```diff
@@ -73,6 +73,35 @@ TEXT runtime·callbackasm(SB),NOSPLIT|NOFRAME,$0
 	}
 }
 
+//todo(ragav): check for correctness
```
```suggestion
//TODO(ragav): check for correctness
```

dmcilvaney commented on src/syscall/types_windows_arm64.go:
```diff
@@ -0,0 +1,22 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package syscall
```
Is this intentionally duplicated from src/cmd/vendor/golang.org/x/sys/windows/types_windows_arm64.go ?

ragavsachdeva replied:
Considering arm32 does the same, yes this is intentional.

hbeberman commented on src/syscall/syscall_windows.go:
```diff
@@ -350,7 +350,8 @@ func setFilePointerEx(handle Handle, distToMove int64, newFilePointer *int64, wh
 		// https://msdn.microsoft.com/en-us/library/windows/desktop/aa383713(v=vs.85).aspx
 		_, _, e1 = Syscall6(procSetFilePointerEx.Addr(), 5, uintptr(handle), uintptr(distToMove), uintptr(distToMove>>32), uintptr(unsafe.Pointer(newFilePointer)), uintptr(whence), 0)
-	case "arm":
+	// Todo(ragav): Verify arm64 case.
```
```suggestion
	// TODO(ragav): Verify arm64 case.
```

hbeberman reviewed (changes requested):
Awesome that hello world is working!

Your indentation is all over the place in the .s files. Looking at others in the src/runtime folder there only ever seems to be a single tab after the instruction.

neilsh-msft commented on src/runtime/asm_arm64.s:
```diff
@@ -15,17 +15,28 @@ TEXT runtime·rt0_go(SB),NOSPLIT,$0
 	MOVD	R7, (g_stack+stack_hi)(g)
 
+#ifdef GOOS_windows
+	BL		runtime·alloc_tls(SB)
```
Too many spaces between BL and runtime alloc_tls?

ragavsachdeva replied:
Fixed.

neilsh-msft commented on src/runtime/syscall_windows.go:
```diff
@@ -38,6 +38,9 @@ func callbackasm()
 // hence 8 bytes.
+// On ARM64, runtime.callbackasm is a series of mov and branch instructions.
+// R16 is loaded with the callback index. Each entry is two instructions,
+// hence 16 bytes. Note(ragav): need to verify this.
```
Note() should be Todo()?

ragavsachdeva replied:
Sure, makes sense.

neilsh-msft commented on src/runtime/sys_windows_arm64.s:
```diff
@@ -0,0 +1,501 @@
+	
+	// Stack must be 16-byte aligned
+	MOVD	RSP, R13
+	BIC		$0xF, R13
```
indenting.

ragavsachdeva replied:
Fixed.

Commit:
Fixed onosstack fn and indentation issues.

ragavsachdeva committed:
Fixed TODOs

Google [style guidelines](https://google.github.io/styleguide/cppguide.html#TODO_Comments) say it should be TODO in all caps.

Co-Authored-By: Henry Beberman <henry.beberman@microsoft.com>

christopherco commented on src/cmd/link/internal/ld/pe.go:
```diff
@@ -1426,11 +1444,12 @@ func addPEBaseRelocSym(ctxt *Link, s *sym.Symbol, rt *peBaseRelocTable) {
 
 func addPEBaseReloc(ctxt *Link) {
-	// We only generate base relocation table for ARM (and ... ARM64), x86, and AMD64 are marked as legacy
+	// We only generate base relocation table for ARM (and ... ARM64). x86 and AMD64 are marked as legacy
```
nit - this looks like an unintended change.

ragavsachdeva replied:
I did intend that. Using a comma after "... ARM64)" was confusing to me. That's meant to be the end of the sentence. Only x86 and AMD64 are marked as legacy systems but the comma makes it misleading. Is this an unnecessary change?

christopherco commented on src/cmd/vendor/golang.org/x/sys/windows/svc/sys_arm64.s:
```diff
@@ -0,0 +1,33 @@
+// func servicemain(argc uint32, argv **uint16)
+TEXT ·servicemain(SB),NOSPLIT|NOFRAME,$0
+    MOVD	R0, ·sArgc(SB)
+ 	MOVD	R1, ·sArgv(SB)
```
nit - fix text alignment

christopherco commented on src/runtime/asm_arm64.s:
```diff
@@ -15,17 +15,28 @@ TEXT runtime·rt0_go(SB),NOSPLIT,$0
-	// _cgo_init may update stackguard.
+	
 	MOVD	$runtime·g0(SB), g
+	// set the per-goroutine and per-mach "registers"
```
Just curious, why was this code moved here from the nocgo label?

ragavsachdeva replied:
The runtime·alloc_tls function (for windows) needs this before being called. My hunch is that moving this here should not affect other architectures but I could very easily be wrong here. 

The alternative is to make a separate pipeline for windows using conditionals and preserve the order of execution for other archs (as is) but it comes at the cost of code repetition. For the purpose of simplicity, I might just do this so I can avoid having to test other archs for now. We can perhaps revisit this later (when upstreaming)? Let me know your thoughts (changed this for now).

Commit:
More indetation fixes and suggested changes as per the pr.

christopherco commented on src/runtime/asm_arm64.s:
```diff
@@ -15,17 +15,28 @@ TEXT runtime·rt0_go(SB),NOSPLIT,$0
+	BL		runtime·alloc_tls(SB)
+#else
 	// if there is a _cgo_init, call it using the gcc ABI.
+	// _cgo_init may update stackguard.
```
nit - move this comment back to its original place to reduce code deltas.

ragavsachdeva replied:
Ok

Commit:
Merge branch 'windows-arm64-port' of https://github.com/ragavsachdeva/go into windows-arm64-port

christopherco commented on src/runtime/tls_arm64.s:
```diff
@@ -26,10 +36,19 @@ nocgo:
 	MOVB	runtime·iscgo(SB), R0
 	CMP	$0, R0
 	BEQ	nocgo
-
```
nit - leave space in

ragavsachdeva replied:
Ok

christopherco commented on src/runtime/tls_arm64.s:
```diff
@@ -38,9 +57,9 @@ TEXT runtime·save_g(SB),NOSPLIT,$0
 	MOVD	runtime·tls_g(SB), R27
 	ADD	R27, R0
 	MOVD	g, 0(R0)
-
```
nit - leave space in

ragavsachdeva replied:
Ok

Commit:
Fixed asm_arm64 to restore execution pipeline for other archs.
